lib/ngtsc/symbol.ts

```diff
@@ -0,0 +1,87 @@
+import { WorkspaceSymbols } from './workspace.symbols';
+import { ClassDeclaration, Decorator } from '@angular/compiler-cli/src/ngtsc/reflection';
+import { Declaration } from 'typescript';
+import { TraitState, Trait } from '@angular/compiler-cli/src/ngtsc/transform';
+import { R3DependencyMetadata } from '@angular/compiler';
+import type { AnnotationNames } from './utils';
+import type { FactoryOutput } from './find-symbol';
+
+const handlerName = {
+  'NgModule': 'NgModuleDecoratorHandler',
+  'Pipe': 'PipeDecoratorHandler',
+  'Injectable': 'InjectableDecoratorHandler',
+  'Directive': 'DirectiveDecoratorHandler',
+  'Component': 'ComponentDecoratorHandler'
+} as const;
+
+const filterByHandler = (annotation: AnnotationNames) => (trait: Trait<Decorator, any, unknown>) => {
+  return trait.handler.name === handlerName[annotation];
+};
+
+
+export abstract class Symbol<AnalysisData> {
+  protected readonly abstract annotation: AnnotationNames;
+  protected readonly abstract deps: R3DependencyMetadata[] | 'invalid' | null;
+  private _trait?: Trait<Decorator, AnalysisData, unknown>;
+
+  constructor(
+    protected workspace: WorkspaceSymbols,
+    public node: ClassDeclaration<Declaration>,
+  ) {}
+
+  get name() {
+    return this.node.name.getText();
+  }
+
+  get diagnostics() {
+    return this.trait?.state === TraitState.ERRORED ? this.trait.diagnostics : null;
+  }
+
+  get isAnalysed() {
+    return this.trait?.state === TraitState.ANALYZED || this.trait?.state === TraitState.RESOLVED;
+  }
+
+  get record() {
+    return this.workspace.traitCompiler?.recordFor(this.node);
+  }
+
+  get analysis(): AnalysisData {
+    this.ensureAnalysis();
+    if (this.trait?.state === TraitState.ERRORED) {
+      const message = `An error occured during analysis of "${this.name}". `;
+      const solution = `Check diagnostics in [${this.annotation}Symbol].diagnostics. `;
+      throw new Error(message + solution);
+    }
+    // As we analyzed the node above it should be ok...
+    if (this.trait?.state === TraitState.ANALYZED || this.trait?.state === TraitState.RESOLVED) {
+      return this.trait.analysis;
+    } else {
+      throw new Error(`Analysis for node ${this.name} couldn't be completed`);
+    }
+  }
+
+  protected get trait() {
+    if (!this._trait) {
+      this._trait = this.record?.traits.find(filterByHandler(this.annotation)) as any;
```

Comment:
If we're sure the type matches we can ascribe:

```suggestion
      this._trait = this.record?.traits.find(filterByHandler(this.annotation)) as Trait<Decorator, AnalysisData, unknown>;
```